adr/2025_03_10-suggestions/README.md

```diff
@@ -0,0 +1,172 @@
+
+- A user can turn on a mode, suggestion mode, which makes all of their edits to a document be tracked, but stored separately from the document itself.
+  - The original document should be unmodified when suggestion mode is turned back off.
+- These stored changes, can be reviewed, and either approved or rejected separately.
+  - Approving a suggestion should apply the change to the document, Rejecting should drop that suggestion
+- A suggestion may become "invalid" if the document that the change has been modified in a way that the change can no longer be applied (e.g. an insertion into a deleted paragraph, or a deletion of an already deleted paragraph)
+- A suggestion should store:
+  - The content of the suggestion based on the type of suggestion
+  - **insertions:** where to insert the change, the content to insert
+  - **deletions:** where to delete the content, the content to be deleted
+  - **updates:** where to update the content, what key to update, what value to set the key to
+  - A timestamp of when the suggestion was proposed
+  - Which user proposed the suggestion
+  - An optional thread identifier, for discussion around why the change was made
+
+**Prior Art:**
+
+- Google Docs: [video here]
+
+### Similarities
+
+Versioning and suggestions have quite a bit of overlap between them, they both rely on having a set of changes which store:
+
+- user: who made the change (the change's attribution)
+- timestamp: of when the change was made
+- content: what changed & it's location in the document
+
+Ideally, we could re-use some of the logic and data structures between them for both features, since they overlap in representation and visualization to users.
+
+## What we need to decide
+
+### How should Y.js snapshots be compared
+
+Y.js already has versioning in the form of snapshots of the Y.js document, at a point in time. In Y.js, the snapshots have the nice property of having some form of a history associated to those edits. To figure out what changed, you could calculate a diff between the two histories of the document. Doing a comparison based on the document content alone, would not meet our requirement of being able to display who made the change and when. Luckily, in Y.js, there is additional metadata that we can extract from Y.js which can tell us which `clientID` made the change (`clientID` being an ID a user assumes when making edits to a document). More on how to calculate that diff here: [Diff two Y.js Snapshots](./diff-yjs-snapshots.md)
+
+This would be the most straightforward way to capture changes across versions of a document. The question then becomes how to display those changes within a prosemirror-based editor, which sort of depends on the approach taken for suggestions.
+
+### How should suggestions be stored
+
+There are a few different mental models we can use to represent or derive these changes:
```

Comment:
can you explain a little bit here what "these changes" are? what makes suggestions different, etc?